File: benchmarks/tsb/bench_clip_advanced.ts
/**
 * Benchmark: clipAdvancedSeries / clipAdvancedDataFrame — per-element clipping with array bounds.
 * Outputs JSON: {"function": "clip_advanced", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, DataFrame, clipAdvancedSeries, clipAdvancedDataFrame } from "../../src/index.ts";

const ROWS = 50_000;
const WARMUP = 5;
const ITERATIONS = 30;

const data = Float64Array.from({ length: ROWS }, (_, i) => Math.sin(i * 0.01) * 200);
const lower = Float64Array.from({ length: ROWS }, () => -50);
const upper = Float64Array.from({ length: ROWS }, () => 50);
const s = new Series(data);
const lowerArr = Array.from(lower);
const upperArr = Array.from(upper);

const dfCols: Record<string, number[]> = {};
for (let c = 0; c < 5; c++) {
  dfCols[`col${c}`] = Array.from({ length: ROWS }, (_, i) => Math.sin((i + c) * 0.01) * 200);
}
const df = new DataFrame(dfCols);

for (let i = 0; i < WARMUP; i++) {
  clipAdvancedSeries(s, { lower: lowerArr, upper: upperArr });
  clipAdvancedDataFrame(df, { lower: -50, upper: 50 });
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  clipAdvancedSeries(s, { lower: lowerArr, upper: upperArr });
  clipAdvancedDataFrame(df, { lower: -50, upper: 50 });
}
const total = performance.now() - start;

console.log(JSON.stringify({ function: "clip_advanced", mean_ms: total / ITERATIONS, iterations: ITERATIONS, total_ms: total }));

File: benchmarks/pandas/bench_clip_advanced.py
"""
Benchmark: Series.clip(lower_arr, upper_arr) / DataFrame.clip() — per-element clipping with array bounds.
Outputs JSON: {"function": "clip_advanced", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import math
import time
import pandas as pd
import numpy as np

ROWS = 50_000
WARMUP = 5
ITERATIONS = 30

data = np.array([math.sin(i * 0.01) * 200 for i in range(ROWS)])
lower = np.full(ROWS, -50.0)
upper = np.full(ROWS, 50.0)
s = pd.Series(data)

df_data = {f"col{c}": np.array([math.sin((i + c) * 0.01) * 200 for i in range(ROWS)]) for c in range(5)}
df = pd.DataFrame(df_data)

for _ in range(WARMUP):
    s.clip(lower=lower, upper=upper)
    df.clip(lower=-50, upper=50)

start = time.perf_counter()
for _ in range(ITERATIONS):
    s.clip(lower=lower, upper=upper)
    df.clip(lower=-50, upper=50)
total = (time.perf_counter() - start) * 1000

print(json.dumps({"function": "clip_advanced", "mean_ms": total / ITERATIONS, "iterations": ITERATIONS, "total_ms": total}))
